runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineResources.java

```diff
@@ -40,34 +44,42 @@
 public class PipelineResources {
 
   /**
-   * Attempts to detect all the resources the class loader has access to. This does not recurse to
-   * class loader parents stopping it from pulling in resources from the system class loader.
+   * Detects all URLs that are present in all class loaders in between context class loader of
+   * calling thread and class loader of class passed as parameter. It doesn't follow parents above
+   * this class loader stopping it from pulling in resources from the system class loader.
    *
-   * @param classLoader The URLClassLoader to use to detect resources to stage.
-   * @throws IllegalArgumentException If either the class loader is not a URLClassLoader or one of
-   *     the resources the class loader exposes is not a file resource.
+   * @param cls Class whose class loader stops recursion into parent loaders
+   * @throws IllegalArgumentException no classloader in context hierarchy is URLClassloader or if
+   *     one of the resources any class loader exposes is not a file resource.
    * @return A list of absolute paths to the resources the class loader uses.
    */
-  public static List<String> detectClassPathResourcesToStage(ClassLoader classLoader) {
-    if (!(classLoader instanceof URLClassLoader)) {
-      String message =
-          String.format(
-              "Unable to use ClassLoader to detect classpath elements. "
-                  + "Current ClassLoader is %s, only URLClassLoaders are supported.",
-              classLoader);
-      throw new IllegalArgumentException(message);
-    }
+  public static List<String> detectClassPathResourcesToStage(Class<?> cls) {
+    return detectClassPathResourcesToStage(cls.getClassLoader());
+  }
 
-    List<String> files = new ArrayList<>();
-    for (URL url : ((URLClassLoader) classLoader).getURLs()) {
-      try {
-        files.add(new File(url.toURI()).getAbsolutePath());
-      } catch (IllegalArgumentException | URISyntaxException e) {
-        String message = String.format("Unable to convert url (%s) to file.", url);
-        throw new IllegalArgumentException(message, e);
-      }
+  /**
+   * Detect resources to stage, by using hierarchy between current context classloader and the given
+   * one. Always include the given class loader in the process.
+   *
+   * @param loader the loader to use as stopping condition when traversing context class loaders
+   * @return A list of absolute paths to the resources the class loader uses.
+   */
+  @VisibleForTesting
+  static List<String> detectClassPathResourcesToStage(final ClassLoader loader) {
+    Set<String> files = new HashSet<>();
```

Comment:
Making this a set even when there is a single classloader means that we will reorder that classloaders classpath which will cause lots of dependency conflict issues.

Reply:
You are right, this is bound to cause trouble. We need to preserve the class loader hierarchy.

Reply:
I can change that to `LinkedHashSet`. But that doesn't solve it entirely, because there is still question in what order to append new files from class loaders. Parent-first? Child-first? What if the class loader of class passed as parameter is not in parent hierarchy of context class loader? Don't see how to define this correctly.